Feature: suggest command usage for misspelt commands

F4zii opened the pull request:
Feature: suggest similar command(s) to a misspelt command

Resolves #742

F4zii commented:
Hey mark, as I've said, the current structure of the error_handler is a bit hard to change for that usage (since it reinvokes every misspelt command as a tag), I can wait or change its implementation if needed
I agree the tags aren't the place for this, I got a suggestion to base it there since its the easiest way, I will now work on a better way and update you :+1: 
I also had an idea to use the returned value from the command invoke into a useful piece of data

F4zii commented:
Hello, the feature is ready.
A quick question:  should we suggest a few options or just the closest one?

Note: Few suggestions could result in a spam

MarkKoz commented on bot/cogs/tags.py:
```diff
@@ -92,9 +92,8 @@ def _get_suggestions(self, tag_name: str, thresholds: Optional[List[int]] = None
-        """Get a specified tag, or a list of all tags if no tag is specified."""
+    async def _get_command(self, ctx: Context, tag_name: TagNameConverter = None) -> None:
+
         def _command_on_cooldown(tag_name: str) -> bool:
```
While you're at it, it'd be nice if this function could get moved out rather than being nested.

F4zii replied:
Are you referring to _command_on_cooldown?

MarkKoz replied:
Yes

MarkKoz commented on bot/cogs/tags.py:
```diff
@@ -92,9 +92,8 @@ def _get_suggestions(self, tag_name: str, thresholds: Optional[List[int]] = None
-    @tags_group.command(name='get', aliases=('show', 'g'))
-    async def get_command(self, ctx: Context, *, tag_name: TagNameConverter = None) -> None:
-        """Get a specified tag, or a list of all tags if no tag is specified."""
+    async def _get_command(self, ctx: Context, tag_name: TagNameConverter = None) -> None:
```
Correct the annotations
```suggestion
    async def _get_command(self, ctx: Context, tag_name: str = None) -> bool:
```

MarkKoz commented on bot/cogs/tags.py:
```diff
@@ -92,9 +92,8 @@ def _get_suggestions(self, tag_name: str, thresholds: Optional[List[int]] = None
-    @tags_group.command(name='get', aliases=('show', 'g'))
-    async def get_command(self, ctx: Context, *, tag_name: TagNameConverter = None) -> None:
-        """Get a specified tag, or a list of all tags if no tag is specified."""
+    async def _get_command(self, ctx: Context, tag_name: TagNameConverter = None) -> None:
```
Considering it is accessed elsewhere (i.e. in `error_handler`), it doesn't seem appropriate to make this function "private" by prefixing its name with an underscore. And since it won't be private, it should have a docstring (explain what it returns too).

F4zii replied:
Alright, right away

F4zii replied:
This is the command itself, are you referring to the explicit one we separated?

F4zii replied:
The outer function (the one that is being called inside the command) is accessible within the error handler, I applied these changes to the outer function

MarkKoz replied:
Was referring to `_get_command` to be clear.

MarkKoz commented on bot/cogs/tags.py:
```diff
@@ -146,6 +147,7 @@ def _command_on_cooldown(tag_name: str) -> bool:
                     description="**There are no tags in the database!**",
                     colour=Colour.red()
                 ))
+                return True
```
It'd be better to return `False` so that it can try to suggest a command instead. The embed saying "there are no tags" would still be visible, but I think it should just not be shown at all if it's being called from the error handler (can add a boolean parameter to the function to distinguish it).

F4zii replied:
That's correct, I think i meant to do this, my bad

MarkKoz replied:
This still needs to be addressed.

F4zii replied:
I changed the implementation to return false, I will check what is wrong.

MarkKoz replied:
I think you are confused about what this comment refers to. It's specifically regarding an embed being sent with the message `"**There are no tags in the database!**"`, which I'm suggesting should _not_ be sent at all if called from the error handler (i.e. merely returning `False` will not be adequate).

F4zii replied:
Oh ok, but we also need to return false at the end of the function, correct?

MarkKoz replied:
Don't think it matters since that line will never be reached (every branch of execution already has a return statement), but I suppose no harm in keeping it.

F4zii replied:
Yeah, correct

MarkKoz replied:
Like I said, 

> merely returning False will not be adequate

The **embed should not be sent** when this function is called by the error handler. I've proposed the solution of adding a parameter to indicate that.

F4zii replied:
Both **Current tags** and **There are no tags in the database!** embeds shouldn't be sent if the error handler invoked the function, correct?

MarkKoz replied:
Sorry, never mind. I just realised those only get sent if `tag_name` is None. The error handler never sends `None` as the name so nothing needs to be changed.

F4zii replied:
Yeah, that makes sense, cool, releasing latest commits and i think we are there

MarkKoz commented on bot/cogs/error_handler.py:
```diff
@@ -87,9 +88,29 @@ def __init__(self, bot: Bot):
 
                 # Return to not raise the exception
-                with contextlib.suppress(ResponseCodeError):
-                    await ctx.invoke(tags_get_command, tag_name=ctx.invoked_with)
```
Is suppressing this exception no longer necessary? Maybe not since the tag suggestion feature, but not sure.

MarkKoz commented on bot/cogs/error_handler.py:
```diff
@@ -87,9 +88,29 @@ def __init__(self, bot: Bot):
                     return
 
                 # Return to not raise the exception
```
This comment may be irrelevant now.

MarkKoz commented on bot/cogs/error_handler.py:
```diff
@@ -87,9 +88,29 @@ def __init__(self, bot: Bot):
                 # Return to not raise the exception
-                with contextlib.suppress(ResponseCodeError):
-                    await ctx.invoke(tags_get_command, tag_name=ctx.invoked_with)
+                tags_cog = self.bot.get_cog("Tags")
```
Should check if it's `None` in case the cog isn't loaded.

MarkKoz commented on bot/cogs/error_handler.py:
```diff
@@ -87,9 +88,29 @@ def __init__(self, bot: Bot):
-                    await ctx.invoke(tags_get_command, tag_name=ctx.invoked_with)
+                tags_cog = self.bot.get_cog("Tags")
+                sent = await tags_cog._get_command(ctx, command_name)
+                # sent = await tags_get_command.callback(tags_get_command.cog, ctx, ctx.invoked_with)
```
This can be removed.

MarkKoz commented on bot/cogs/error_handler.py:
```diff
@@ -87,9 +88,29 @@ def __init__(self, bot: Bot):
+                raw_commands = [
+                    (cmd.name, *cmd.aliases)
+                    for cmd in self.bot.walk_commands()
+                    if not cmd.hidden
```
Can't these two comprehensions be combined into one? What even is the purpose of the second one? Flattening the list?

F4zii replied:
Yeah, from List[Tuple] (command_name, aliases...) into a list of command names and aliases

MarkKoz replied:
Wouldn't this single loop be equivalent to the two list comprehensions?

```py
raw_commands = []
for cmd in self.bot.walk_commands():
    if not cmd.hidden:
        raw_commands += (cmd.name, *cmd.aliases)
```

MarkKoz commented on bot/cogs/error_handler.py:
```diff
@@ -87,9 +88,29 @@ def __init__(self, bot: Bot):
+                if similar_command.can_run(ctx):
+                    misspelled_content = ctx.message.content
+                    await ctx.send(
+                        f"Did you mean:\n**{misspelled_content.replace(command_name, similar_command_name)}**",
```
This could mistakenly replace something other than the command name too. Should use a more precise method of replacing which only affects the command name.

MarkKoz commented on bot/cogs/error_handler.py:
```diff
@@ -87,9 +88,29 @@ def __init__(self, bot: Bot):
+                    for cmd in self.bot.walk_commands()
+                    if not cmd.hidden
+                ]
+                raw_commands = [c for data in raw_commands for c in data]
```
Perhaps this could be cached but it seems a bit complicated to keep it up to date considering commands can be added/removed during runtime (e.g. via unloading extensions). Not sure if worth looking into.

F4zii replied:
I will note it :ok_hand:

MarkKoz commented on bot/cogs/error_handler.py:
```diff
@@ -87,9 +88,29 @@ def __init__(self, bot: Bot):
+                similar_command_data = difflib.get_close_matches(command_name, raw_commands, 1)
+                similar_command_name = similar_command_data[0]
+                similar_command = self.bot.get_command(similar_command_name)
+                if similar_command.can_run(ctx):
```
This will raise exceptions for check failures so they should be caught.

MarkKoz commented on bot/cogs/tags.py:
```diff
@@ -92,7 +92,18 @@ def _get_suggestions(self, tag_name: str, thresholds: Optional[List[int]] = None
+
+        Returns False if no tag is not found or the tag is on cooldown
+
+        Return True if found and sent a tag
```
Keeping in mind this is not strictly enforced, here are my suggestions for writing docstrings:

1. Use the imperative mood.
2. Try to include information about the return value in the summary, if it fits.
3. A parameters/arguments section is generally something we don't use. If they warrant explicit explanation, then that information should be provided as sentences/paragraphs in the body of the docstring.

My attempt, feel free to modify:

```py
        """
        Show contents of the tag `tag_name` in `ctx` and return True if something is shown.

        If a tag is not found, display similar tag names as suggestions. If a tag is not specified,
        display a paginated embed of all tags.

        Tags are on cooldowns on a per-tag, per-channel basis. If a tag is on cooldown, display
        nothing and return False.
        """
```

MarkKoz commented on bot/cogs/tags.py:
```diff
@@ -92,7 +92,18 @@ def _get_suggestions(self, tag_name: str, thresholds: Optional[List[int]] = None
         await ctx.invoke(self.get_command, tag_name=tag_name)
 
-    async def _get_command(self, ctx: Context, tag_name: TagNameConverter = None) -> None:
+    async def get_command(self, ctx: Context, tag_name: str = None) -> bool:
```
This isn't a command so it should not be named as such. `display_tag` is better, for starters.

MarkKoz commented on bot/cogs/tags.py:
```diff
@@ -160,12 +171,12 @@ def _command_on_cooldown(tag_name: str) -> bool:
-    async def get_command(self, ctx: Context, *, tag_name: TagNameConverter = None) -> None:
-        """Get a specified tag, or a list of all tags if no tag is specified."""
-        await self._get_command(ctx, tag_name)
+    async def _get_command(self, ctx: Context, *, tag_name: TagNameConverter = None) -> None:
```
This name could have just been left alone since, as I mentioned, `get_commands` is not an appropriate name for the other function anyway.

MarkKoz commented on bot/cogs/error_handler.py:
```diff
@@ -86,10 +92,33 @@ def __init__(self, bot: Bot):
+                        await ctx.send(
+                            embed=e,
+                            delete_after=7.0
+                        )
```
This doesn't need to be broken up into multiple lines.

MarkKoz commented on bot/cogs/error_handler.py:
```diff
@@ -108,16 +108,20 @@ def __init__(self, bot: Bot):
+                        return
+                except CommandError as cmd_error:
+                    log.debug(log_msg)
+                    await self.on_command_error(ctx, cmd_error)
```
Need to return after this otherwise the embed still gets sent.

MarkKoz commented on bot/cogs/error_handler.py:
```diff
@@ -86,10 +92,36 @@ def __init__(self, bot: Bot):
+
+                log_msg = "Cancelling attempt to suggest a command due to failed checks."
+                try:
+                    if not similar_command.can_run(ctx):
```
`can_run` needs to be awaited.

MarkKoz commented on bot/cogs/tags.py:
```diff
@@ -146,7 +146,7 @@ def command_on_cooldown(self, ctx: Context, tag_name: str) -> bool:
                     description='\n'.join(tag['title'] for tag in founds[:10])
                 ))
-                return True
+                return False
```
This should still return True otherwise the bot will send a command suggestion after the tag suggestions, which is not desirable.

F4zii replied:
Yeah, I will add a parameter in that case

ikuyarihS replied:
I concur that this should return a True, otherwise this is possible

![image](https://user-images.githubusercontent.com/19921765/75602909-29564700-5afc-11ea-8a21-4b7f1cedeec8.png)

MarkKoz commented:
I don't know how you merged but you did something wrong cause all the commits are co-authored by you and that should absolutely not happen. Do you know how to fix this? If not, I can attempt a fix myself.

F4zii commented:
> I don't know how you merged but you did something wrong cause all the commits are co-authored by you and that should absolutely not happen. Do you know how to fix this? If not, I can attempt a fix myself.

I think it'll be better if you try, I just meged master into my branch, weird

MarkKoz commented:
Okay. Please don't push anything until I let you know I've fixed it.

MarkKoz commented:
OK should be good to go. By the way, there was a lint error - `contextlib` was imported but unused. I fixed it. Do you not have the pre-commit hook installed?

F4zii commented:
> OK should be good to go. By the way, there was a lint error - `contextlib` was imported but unused. I fixed it. Do you not have the pre-commit hook installed?

Yeah I fixed it already, maybe the merge / pull changed it, thanks for fixing it

MarkKoz commented on bot/cogs/error_handler.py:
```diff
@@ -74,9 +75,13 @@ def __init__(self, bot: Bot):
+                tags_cog = self.bot.get_cog("Tags")
                 tags_get_command = self.bot.get_command("tags get")
-                ctx.invoked_from_error_handler = True
+                if not tags_cog and not tags_get_command:
```
```suggestion
                if not tags_cog or not tags_get_command:
```

F4zii replied:
Good catch, fixing...

ikuyarihS commented:
Right now there is a rare chance that it can be a tag instead, can we move it up to check for commands, specially if invoker is a mod+?

![image](https://user-images.githubusercontent.com/19921765/75602636-c95ea100-5af9-11ea-9a3f-8a3353058717.png)

F4zii commented:
> Right now there is a rare chance that it can be a tag instead, can we move it up to check for commands, specially if invoker is a mod+?
> 
> ![image](https://user-images.githubusercontent.com/19921765/75602636-c95ea100-5af9-11ea-9a3f-8a3353058717.png)

@ikuyarihS, I think it's enough to check if the Command was invoked with arguments since if that's a single tag, no following arguments will be provided

lemonsaurus commented:
> @ikuyarihS, I think it's enough to check if the Command was invoked with arguments since if that's a single tag, no following arguments will be provided

I'm not sure I buy this line of reasoning. some moderation commands (like `!silence`) do not take arguments. I think prefering commands over tags is probably the right suggestion.

@F4zii Do you have time to get this wrapped up? It looks like it's fallen victim to Covid19 and we should try to nurse it back to health and get it merged.

Xithrius commented:
I screwed up, I am sorry.

Xithrius commented:
Fixed.

Xithrius commented:
Works with long commands and their arguments.
![image](https://user-images.githubusercontent.com/15021300/103471971-30616d80-4d3c-11eb-83b7-49da77bc7a95.png)

Works with the smallest commands as well.
![image](https://user-images.githubusercontent.com/15021300/103471979-57b83a80-4d3c-11eb-9cf7-0edcc1d46872.png)

So far I haven't found any bugs with this command and the way that it's implemented.

Still deciding if this should exist within mod channels at all, since it may just be noise at that point to the mods. At the same time, the error stays up for 10 seconds, so instead of having to retype/edit the command, it can be copy/pasted.

Xithrius commented:
A user with no roles executing a moderator+ only command along with a regular command (typed incorrectly) that anyone can execute.
![image](https://user-images.githubusercontent.com/15021300/103733481-aebd4a00-4f9e-11eb-87b5-e12a1bd7487c.png)

Xithrius commented:
This is the outcome of a recent poll. I think the conclusion that I can reach here is that only helpers and below should get command suggestions.
![image](https://user-images.githubusercontent.com/15021300/103838557-39a35080-5042-11eb-86fc-3922856b375e.png)

Xithrius commented:
Here are examples of a user at different roles executing commands incorrectly.

No-role user:
![image](https://user-images.githubusercontent.com/15021300/103841447-76724600-5048-11eb-9c34-87531e7fa77e.png)


Helper:
![image](https://user-images.githubusercontent.com/15021300/103841409-665a6680-5048-11eb-834f-1090d61eabe4.png)

Moderator:
![image](https://user-images.githubusercontent.com/15021300/103841463-7eca8100-5048-11eb-9438-712f49d47271.png)

Admin:
![image](https://user-images.githubusercontent.com/15021300/103841531-a9b4d500-5048-11eb-9135-c8def0c814c3.png)

jerbob commented:
![](https://cdn.discordapp.com/emojis/750350748746907669.gif?v=1)
